machine-check/src/machine.rs

```diff
@@ -0,0 +1,367 @@
+            }
+        }
+    }
+
+    pub fn convert<'a>(left: &'a Graph, right: &'a Graph) -> impl Fn(Error) -> String + 'a {
+        |err| err.to_string(left, right)
+    }
+}
+
+struct StatePrinter<'a>(Option<&'a State>, u32);
+
+impl<'a> fmt::Display for StatePrinter<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.0.map(|s| s.as_ref()).unwrap_or("[invalid]"))?;
+        if self.1 > 0 {
+            write!(f, "(+{})", self.1)?;
+        }
+        Ok(())
+    }
+}
+
+fn state_name(g: &Graph, mut n: NodeId) -> StatePrinter<'_> {
+    let mut offset = 0;
+    loop {
+        match g.node_weight(n) {
+            Some(Some(state)) => return StatePrinter(Some(state), offset),
+            None => return StatePrinter(None, offset),
+            _ => {}
+        }
+        tracing::debug!(?n, "tracking back");
+        n = g
+            .neighbors_directed(n, Incoming)
+            .next()
+            .expect("unnamed state must track back to named state");
+        offset += 1;
+    }
+}
+
+/// error messages are designed assuming that `left` is the reference and `right` the tested
+pub fn equivalent(left: &Graph, li: NodeId, right: &Graph, ri: NodeId) -> Vec<Error> {
+    use Side::*;
+
+    let _span = tracing::debug_span!("equivalent").entered();
+
+    let mut errors = Vec::new();
+    let mut l2r = vec![NodeId::end(); left.node_count()];
+    let mut r2l = vec![NodeId::end(); right.node_count()];
+
+    // dfs traversal stack
+    // must hold index pairs because node mappings might be m:n
+    let mut stack = vec![(li, ri)];
+
+    while let Some((li, ri)) = stack.pop() {
+        tracing::debug!(left = %state_name(left, li), ?li, right = %state_name(right, ri), ?ri, to_go = stack.len(), "loop");
+        // get all outgoing edge labels for the left side
+        let mut l_out = BTreeMap::new();
+        for edge in left.edges_directed(li, Outgoing) {
+            l_out
+                .entry(edge.weight())
+                .and_modify(|_| errors.push(Error::NonDeterministic(Left, edge.id())))
+                .or_insert(edge);
+        }
+        // get all outgoing edge labels for the right side
+        let mut r_out = BTreeMap::new();
+        for edge in right.edges_directed(ri, Outgoing) {
+            r_out
+                .entry(edge.weight())
+                .and_modify(|_| errors.push(Error::NonDeterministic(Right, edge.id())))
+                .or_insert(edge);
+        }
+        // keep note of stack so we can undo additions if !same
+        let stack_len = stack.len();
+        // note that we have visited these nodes (to avoid putting self-loops onto the stack in the loop below)
+        l2r[li.index()] = ri;
+        r2l[ri.index()] = li;
+        // compare both sets
+        let mut same = true;
+        let mut l_edges = l_out.into_iter().peekable();
+        let mut r_edges = r_out.into_iter().peekable();
+        loop {
```

Comment:
This loop seems to be relying that edges for both `l` and `r` are ordered based on their content (which I assume is provided by `derive(Ord)`. Does inserting edges as a key of BTreeMap and then taking it out via `into_iter` automatically order it based on its content?